SonarQube.Client.Tests/Helpers/SecondaryIssueHashUpdaterTests.cs

```diff
@@ -0,0 +1,161 @@
+                    AddFlow(
+                        CreateLocation("duplicate"),
+                        CreateLocation("duplicate"),
+                        CreateLocation("unique1"),
+                        CreateLocation("unique2")
+                        )),
+                CreateIssue("unique2",
+                    AddFlow(
+                        CreateLocation("unique3")
+                        )),
+                CreateIssue("primary_only_should_not_be_fetched_2",
+                    AddFlow(
+                        CreateLocation("unique4")
+                        ))
+            };
+
+            // Only expecting the unique set of secondary locations to be requested
+            var serviceMock = new Mock<ISonarQubeService>();
+            AddSourceFile(serviceMock, "duplicate");
+            AddSourceFile(serviceMock, "unique1");
+            AddSourceFile(serviceMock, "unique2");
+            AddSourceFile(serviceMock, "unique3");
+            AddSourceFile(serviceMock, "unique4");
+
+            var testSubject = new SecondaryIssueHashUpdater();
+            await testSubject.UpdateHashesAsync(issues, serviceMock.Object, CancellationToken.None);
+
+            serviceMock.VerifyAll();
+            serviceMock.VerifyNoOtherCalls();
+        }
+
+        [TestMethod]
+        public async Task Populate_HasSecondaryLocations_ExpectedHashesSet()
+        {
+            const string line1Contents = "line one contents";
+            const string line2Contents = " line two XXX ";
+            const string line3Contents = "  LINE THREE!\"£$% ";
+
+            var file1Contents = $"{line1Contents}\n{line2Contents}\nfoo foo foo";
+
+            var file2Contents = $"111\n222\n{line3Contents}";
+
+            var issues = new[]
+            {
+                CreateIssue("primary_only_should_not_be_fetched_1",
+                    AddFlow(
+                        CreateLocation("file1", startLine: 1),
+                        CreateLocation("file1", startLine: 2)
+                        )),
+                CreateIssue("primary_only_should_not_be_fetched_2",
+                    AddFlow(
+                        CreateLocation("file2", startLine: 4), // beyond the end of the file -> should be ignored
+                        CreateLocation("file2", startLine: 3)
+                        ))
+            };
+
+            var serviceMock = new Mock<ISonarQubeService>();
+            AddSourceFile(serviceMock, "file1", file1Contents);
+            AddSourceFile(serviceMock, "file2", file2Contents);
+
+            var testSubject = new SecondaryIssueHashUpdater();
+
+            // Act
+            await testSubject.UpdateHashesAsync(issues, serviceMock.Object, CancellationToken.None);
+
+            issues[0].Hash.Should().Be(PrimaryIssueHash);
+            issues[1].Hash.Should().Be(PrimaryIssueHash);
+
+            issues[0].Flows[0].Locations[0].Hash.Should().Be(checksumCalculator.Calculate(line1Contents));
```

Comment:
It's already an interface, why not mock it? its logic is irrelevant to this class, it shouldn't be a real dependency.
It can be easily replaced with a setup method:
```
SetupHash(checksumCalculator, line1Contents, "expected hash1");
SetupHash(checksumCalculator, line2Contents, "expected hash2");
...
issues[0].Flows[0].Locations[0].Hash.Should().Be("expected hash1");
issues[0].Flows[0].Locations[1].Hash.Should().Be("expected hash2");

private void SetupHash(Mock<IChecksumCalculator> checksumCalculator, string originalContent, string hashedContent) {...}
```